runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java

```diff
@@ -592,12 +593,37 @@ public long applyOutputWatermarkHold(long currentOutputWatermark, long potential
         return currentOutputWatermark;
       }
     } else {
+      // Check if the final watermark was triggered to perform state cleanup for global window
+      if (potentialOutputWatermark == Long.MAX_VALUE
+          && currentOutputWatermark < Long.MAX_VALUE
+          && isStateful) {
+        cleanupGlobalWindowState();
+      }
```

Comment:
```suggestion
      // Check if the final watermark was triggered to perform state cleanup for global window
      if (potentialOutputWatermark > BoundedWindow.maxTimestamp().getMillis()
          && currentOutputWatermark <= BoundedWindow.maxTimestamp().getMillis()) {
        cleanupGlobalWindowState();
      }
```

Would change the condition to specifically target the end of the global window. There is also no need to include the statefulness in the condition because if there is no state, it also won't be cleaned up.